plugin/src/main/java/io/jenkins/plugins/coverage/model/CodeDeltaCalculator.java

```diff
@@ -0,0 +1,111 @@
+import io.jenkins.plugins.forensics.delta.model.Delta;
+import io.jenkins.plugins.forensics.delta.model.FileChanges;
+import io.jenkins.plugins.forensics.delta.model.FileEditType;
+import io.jenkins.plugins.forensics.git.delta.GitDeltaCalculatorFactory;
+
+public class CodeDeltaCalculator {
+
+    private final Run<?, ?> build;
+    private final FilePath workspace;
+    private final TaskListener listener;
+
+    public CodeDeltaCalculator(@NonNull final Run<?, ?> build, @NonNull final FilePath workspace,
+            @NonNull final TaskListener listener) {
+        this.build = build;
+        this.workspace = workspace;
+        this.listener = listener;
+    }
+
+    public void calculateCodeDeltaInTree(final Run<?, ?> referenceBuild, final CoverageNode rootNode,
+            final FilteredLog log) {
+        BuildData buildAction = build.getAction(BuildData.class);
+        BuildData previousBuildAction = referenceBuild.getAction(BuildData.class);
+
+        if (buildAction != null && previousBuildAction != null) {
+            String commit = buildAction.lastBuild.getRevision().getSha1String();
+            String previousCommit = previousBuildAction.lastBuild.getRevision().getSha1String();
+
+            DeltaCalculator deltaCalculator = getDeltaCalculator();
+            Optional<Delta> delta = deltaCalculator.calculateDelta(commit, previousCommit, log);
+
+            delta.ifPresent(value -> setChangedFilesInCoverageTree(rootNode, value));
+        }
+    }
+
+    private DeltaCalculator getDeltaCalculator() {
+        FilteredLog log = new FilteredLog("Calculate code delta");
+        SCM scm = ((AbstractBuild<?, ?>) build).getProject().getScm();
+        if (scm instanceof GitSCM) {
+            GitDeltaCalculatorFactory gitDeltaCalculatorFactory = new GitDeltaCalculatorFactory();
+            Optional<DeltaCalculator> gitDeltaCalculator = gitDeltaCalculatorFactory.createDeltaCalculator(scm, build,
+                    workspace, listener, log);
+            if (gitDeltaCalculator.isPresent()) {
+                return gitDeltaCalculator.get();
+            }
+        }
+        return DeltaCalculatorFactory
+                .findDeltaCalculator(build, Collections.singleton(workspace), listener, log);
+    }
+
+    private void setChangedFilesInCoverageTree(final CoverageNode coverageNode, final Delta delta) {
+        Map<String, CoverageNode> nodePathMapping = coverageNode.getAll(CoverageMetric.FILE).stream()
+                .map(node -> (FileCoverageNode) node)
+                .collect(Collectors.toMap(FileCoverageNode::getPath, Function.identity()));
+
+        List<FileChanges> fileChanges = delta.getFileChangesMap().values().stream()
+                .filter(fileChange -> fileChange.getFileEditType().equals(FileEditType.MODIFY)
+                        || fileChange.getFileEditType().equals(FileEditType.ADD))
+                .collect(Collectors.toList());
+
+        for (FileChanges fileChange : fileChanges) {
```

Comment:
Warum erst in eine Liste packen wenn darüber dann nur iteriert wird? .forEach(fileChange->{...})